### PR TITLE
Add azure provider

### PR DESCRIPTION
**Pull request by anvddriesch**
Towards https://github.com/giantswarm/roadmap/issues/1432 and https://github.com/giantswarm/giantswarm/issues/21627

This adds the azure provider to the operator.
It implements the `provider` interface and creates and deletes app configuration on the azure active directory provider if credentials are given.
No update logic is included yet.

### CODE REVIEW

File: helm/dex-operator/templates/secret.yaml [hunk 1]
```diff
@@ -1,15 +1,26 @@
 apiVersion: v1
 stringData:
   credentials: |-
-    - name: mock
+    {{- if .Values.oidc.customer.providers }}
```
Comment (anvddriesch) [inline helm/dex-operator/templates/secret.yaml:4]: We allow an array of identity provider credentials for both giantswarm and customers

File: pkg/idp/idp.go [hunk 2]
```diff
@@ -131,33 +132,41 @@ func (s *Service) ReconcileDelete(ctx context.Context) error {
...
-		connector, err := provider.CreateApp(appConfig)
+		connector, err := provider.CreateApp(appConfig, ctx)
 		if err != nil {
 			return dexConfig, err
 		}
-
 		// Add connector configuration to config
-		dexConfig.Oidc.Giantswarm.Connectors = append(dexConfig.Oidc.Giantswarm.Connectors, connector)
-
+		switch provider.GetOwner() {
```
Comment (anvddriesch) [inline pkg/idp/idp.go:150]: We write back the connectors according to owner

File: pkg/idp/idp.go [hunk 3]
```diff
@@ -183,8 +192,9 @@ func (s *Service) GetAppConfig(ctx context.Context) (provider.AppConfig, error)
 		}
 	}
 	return provider.AppConfig{
-		Name:        key.GetIdpAppName(s.app.Namespace, s.app.Name),
-		RedirectURI: key.GetRedirectURI(baseDomain),
+		Name:          key.GetIdpAppName(s.app.Namespace, s.app.Name),
+		RedirectURI:   key.GetRedirectURI(baseDomain),
+		IdentifierURI: key.GetIdentifierURI(baseDomain),
```
Comment (anvddriesch) [inline pkg/idp/idp.go:197]: Adding the identifier URI to the app config

File: pkg/idp/provider/azure/azure.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+func (a *Azure) GetType() string {
+	return a.Type
+}
+
+func (a *Azure) GetOwner() string {
+	return a.Owner
+}
+
+func (a *Azure) CreateApp(config provider.AppConfig, ctx context.Context) (dex.Connector, error) {
```
Comment (anvddriesch) [inline pkg/idp/provider/azure/azure.go:82]: This function creates the app on the azure provider

File: pkg/idp/provider/azure/azure.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+	createdApp, err := a.Client.Applications().Post(ctx, getAppCreateRequestBody(config), nil)
+	if err != nil {
+		return dex.Connector{}, microerror.Maskf(requestFailedError, printOdataError(err))
+	}
+	id := createdApp.GetId()
+	if id == nil {
+		return dex.Connector{}, microerror.Maskf(notFoundError, "Could not find ID of app %s.", config.Name)
+	}
+	createdSecret, err := a.Client.ApplicationsById(*id).AddPassword().Post(ctx, getSecretCreateRequestBody(config), nil)
```
Comment (anvddriesch) [inline pkg/idp/provider/azure/azure.go:91]: Adding the credential secret

File: pkg/idp/provider/azure/azure.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+	if err != nil {
+		return dex.Connector{}, microerror.Maskf(requestFailedError, printOdataError(err))
+	}
+	parentApp, err := a.GetApp(DefaultName)
+	if err != nil {
+		return dex.Connector{}, microerror.Mask(err)
+	}
+	_, err = a.Client.ApplicationsById(*id).Patch(ctx, getPermissionCreateRequestBody(parentApp), nil)
```
Comment (anvddriesch) [inline pkg/idp/provider/azure/azure.go:99]: Adding the permissions based on permissions set on a parent app.
The reason we do it that way is that querying permissions by scope (human readable) is not possible in the microsoft graph API and hardcoding the IDs seems risky

File: pkg/idp/provider/azure/azure.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+	if err != nil {
+		return dex.Connector{}, microerror.Maskf(requestFailedError, printOdataError(err))
+	}
+	clientSecret := createdSecret.GetSecretText()
+	if clientSecret == nil {
+		return dex.Connector{}, microerror.Maskf(notFoundError, "Could not find client secret for app %s.", config.Name)
+	}
+	return dex.Connector{
```
Comment (anvddriesch) [inline pkg/idp/provider/azure/azure.go:107]: Creating the connector based on the app that we created

File: pkg/idp/provider/azure/azure.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+		Config: &microsoft.Config{
+			ClientID:     *id,
+			ClientSecret: *clientSecret,
+			RedirectURI:  config.RedirectURI,
+			Tenant:       a.TenantID,
+		},
+	}, nil
+}
+
+func (a *Azure) DeleteApp(name string, ctx context.Context) error {
```
Comment (anvddriesch) [inline pkg/idp/provider/azure/azure.go:120]: Deleting the app by its unique name

File: pkg/idp/provider/azure/errors.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+var requestFailedError = &microerror.Error{
+	Kind: "requestFailedError",
+}
+
+// IsRequestFailed asserts requestFailedError.
+func IsRequestFailed(err error) bool {
+	return microerror.Cause(err) == requestFailedError
+}
+
+func printOdataError(err error) string {
```
Comment (anvddriesch) [inline pkg/idp/provider/azure/errors.go:37]: Helper function to display more information when printing the microsoft graph api errors

File: pkg/idp/provider/azure/request.go [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+	DefaultName           = "giantswarm-dex"
+	Claim                 = "groups"
+	Audience              = "AzureADMyOrg"
+)
+
+func ProviderScope() []string {
+	return []string{"https://graph.microsoft.com/.default"}
+}
+
+func getPermissionCreateRequestBody(parentApp models.Applicationable) models.Applicationable {
```
Comment (anvddriesch) [inline pkg/idp/provider/azure/request.go:30]: To set permissions based on the permissions set in another app

File: pkg/idp/provider/azure/request.go [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+	app := models.NewApplication()
+	app.SetRequiredResourceAccess(parentApp.GetRequiredResourceAccess())
+	return app
+}
+
+func getAppGetRequestConfig(name string) *applications.ApplicationsRequestBuilderGetRequestConfiguration {
```
Comment (anvddriesch) [inline pkg/idp/provider/azure/request.go:36]: To query an app by name

File: pkg/idp/provider/azure/request.go [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+	return app
+}
+
+func getSecretCreateRequestBody(config provider.AppConfig) *addpassword.AddPasswordPostRequestBody {
+	keyCredential := models.NewPasswordCredential()
+	keyCredential.SetDisplayName(&config.Name)
+
+	validUntil := time.Now().AddDate(0, key.SecretValidityMonths, 0)
+	keyCredential.SetEndDateTime(&validUntil)
```
Comment (anvddriesch) [inline pkg/idp/provider/azure/request.go:89]: We set the secret validity here

**Review by vvondruska — APPROVED** [review]
Makes sense and looks good to me

**Pull request merged**